### PR TITLE
drivers: adc: stm32: dma support

### PR DESCRIPTION
**Pull request by heinwessels**
The STM32 ADC already supports sampling multiple channels in one sequence. This nwo PR allows those channels to be read in one go using the DMA.

Support for multiple channel sampling for the STM32 was originally part of this PR, but split off in #53534.

Currently dependent on #54156 for a final check

**Notes:**
-  Currently only tested with STM32H743zi. With only this PR only ADC1 is supported. This PR #53184 enables the `bdma` which will add support for running ADC3 through DMA as well.
- This PR is porting ADC DMA support from a Zephyr 2.6 fork, which has been running on many devices for combined hundreds of hours with the MCUs under heavy load.
- Does not support DMA driven continious sampling, and is considered out-of-scope for this PR because it can already be mimicked with `ADC_ACTION_REPEAT`. Adding continous sampling will require changes to `adc_context`, which has an open discussion here: https://github.com/zephyrproject-rtos/zephyr/discussions/48478

### CODE REVIEW

**Comment by erwango** [comment]
@heinwessels Thanks for your work on this feature which has been requested since a long time by numerous users.
In order to help the review process, please split this PR in few commits, at least:
- One commit dedicated to the sample changes
- One commit dedicated to the changes that impacts the driver when `CONFIG_ADC_STM32_DMA` is not enabled
- One commit for the pure driver DMA related changes

Also, one general comment the solution:
Generally, on other DMA enabled STM32 devices, DMA is enabled based on the presence of DMA properties on the node (coupled to an FOO_ASYNC Kconfig driver symbol if provided by the API). For consistency, it would be great to add that as well.
This is not blocking though, it can be added later on.

**Comment by heinwessels** [comment]
@erwango 
> In order to help the review process, please split this PR in few commits, at least:

Done. Please let me know if you had something different in mind.

> Also, one general comment the solution: Generally, on other DMA enabled STM32 devices, DMA is enabled based on the presence of DMA properties on the node (coupled to an FOO_ASYNC Kconfig driver symbol if provided by the API). For consistency, it would be great to add that as well. This is not blocking though, it can be added later on.

I would prefer to keep it as-is for now, but if it's really required I can work it in.

**Review by erwango — CHANGES_REQUESTED** [review]
Some preliminary changes. Will need more time to dig into adc specifics

File: samples/drivers/adc/boards/nucleo_h743zi.overlay
Comment (heinwessels) [reply samples/drivers/adc/boards/nucleo_h743zi.overlay]: True. Removed.

File: samples/drivers/adc/custom-sections.ld
Comment (heinwessels) [reply samples/drivers/adc/custom-sections.ld]: Cool, I didn't know that file existed. Done.

File: drivers/adc/adc_stm32.c
Comment (FRASTM) [inline drivers/adc/adc_stm32.c:724]: change that are not due to DMA, should better come in another PR

File: drivers/adc/adc_stm32.c
Comment (FRASTM) [inline drivers/adc/adc_stm32.c:922]: _else_ we have no channel value, is it sure that at least one `data->channels & BIT(channel_id))` will be true ?

File: drivers/adc/Kconfig.stm32
Comment (erwango) [inline drivers/adc/Kconfig.stm32:33]: For now, this also depends on DMAMUX

File: drivers/adc/adc_stm32.c
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:724]: As requested by @erwango these changes are part of a seperate commit to which it's relevant https://github.com/zephyrproject-rtos/zephyr/pull/52965/commits/3755487ba43efb083711d0d4e69853335183acf3. I would prefer to keep it as is, but if it's strictly necessary I can move it.

File: drivers/adc/adc_stm32.c
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:632]: As requested by @erwango these changes are part of a seperate commit to which it's relevant https://github.com/zephyrproject-rtos/zephyr/pull/52965/commits/3755487ba43efb083711d0d4e69853335183acf3. I would prefer to keep it as is, but if it's strictly necessary I can move it to a different PR.

File: drivers/adc/adc_stm32.c
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:922]: It's checked after the `for` loop if there were channels selected or not.
```C
	if (data->channel_count == 0) {
		LOG_ERR("No channels selected");
		return -EINVAL;
	}
```

File: drivers/adc/adc_stm32.c
Comment (erwango) [reply drivers/adc/adc_stm32.c:724]: If it can make the whole set being merged faster, let's split in a separate PR.
When doing so, please also provide the related sample changes so the multiple channel feature can be tested independently from DMA. This is indeed some more work to do, but a good thing overall.

Comment (heinwessels) [reply drivers/adc/adc_stm32.c]: Split the PR off to #53534, meaning this PR won't build until that PR is merged. It was designed with multiple channels in mind, and I would strongly prefer not doing redudnent work getting it to work standalone.

File: drivers/adc/adc_stm32.c
Comment (heinwessels) [reply drivers/adc/adc_stm32.c]: Moved to different PR.

**Comment by heinwessels** [comment]
@erwango  @FRASTM 

This PR is ready for review again now that multiple-channel-sampling is supported.

I also added unit tests for the ADC DMA driver.

**Comment by erwango** [comment]
Adding NXP maintainers for the NXP boards related change

**Review by FRASTM — COMMENTED** [review]
sounds good, 
did you also check for stm32 mcu without dmamux ?

**Comment by heinwessels** [comment]
> did you also check for stm32 mcu without dmamux ?

No, but from the ADC drivers point of view there should be no difference between using a DMA and DMAMUX. I will remove accidental  dependency of `ADC_STM32_DMA` on `DT_HAS_ST_STM32_DMAMUX_ENABLED`. It should depend on `STM32_DMA`.

That said, I haven't tested running the ADC driver through `dma1` (and not `dmamux1`) because it doesn't compile IIRC, even on `main`.

**Comment by gautierg-st** [comment]
@heinwessels Could you please rebase this PR? There is a conflict now that #54038 has been merged.

**Review by FRASTM — COMMENTED** [review]
could you detail the reason for adding the section .ld ?
else LGTM

File: include/zephyr/toolchain/common.h
Comment (erwango) [inline include/zephyr/toolchain/common.h:194]: Please put this in a dedicated PR

Comment (heinwessels) [reply include/zephyr/toolchain/common.h:194]: Split of in #54156

**Comment by heinwessels** [comment]
Rebased with the recent driver changes.

> could you detail the reason for adding the section .ld ? else LGTM

I initially thought that `dma1` only had access to `SRAM1`, but realized that this is not the case. So I removed this change.

> Not explicitly, but from the ADC drivers point of view there should be no difference between using a DMA and DMAMUX. I will remove accidental dependency of `ADC_STM32_DMA` on `DT_HAS_ST_STM32_DMAMUX_ENABLED`. It should depend on `STM32_DMA`.

I removed these dependencies because enabling `ADC_STM32_DMA` should do a `select DMA`. It cannot be dependent on `DMA` as well.

Finally, the check to see if the buffer is correctly aligned was also broken. It was checking for alignment for `uint32_t` (32 bits) and not 32 bytes. I fixed this, but this is now dependent on a split-off PR #54156. Depending on the outcome of that PR I will re-enable this check or remove it.

**Comment by heinwessels** [comment]
> Finally, the check to see if the buffer is correctly aligned was also broken. It was checking for alignment for `uint32_t` (32 bits) and not 32 bytes. I fixed this, but this is now dependent on a split-off PR #54156. Depending on the outcome of that PR I will re-enable this check or remove it.

Updated the code because the mentioned PR was merged.

File: drivers/adc/adc_stm32.c
Comment (weinholt) [inline drivers/adc/adc_stm32.c:848]: Is this check only for making sure that `SCB_InvalidateDCache_by_Addr()` gets a 32-byte aligned address? The version of that function used in the current CMSIS properly handles unaligned addresses: https://github.com/zephyrproject-rtos/cmsis/blob/master/CMSIS/Core/Include/cachel1_armv7.h#L320 *D-Cache memory blocks which are part of given address + given size are cleaned.*

There are other variants of the documentation for that function floating about which claim it needs 32-byte alignment, but the one used in Zephyr doesn't.

File: drivers/adc/adc_stm32.c
Comment (weinholt) [inline drivers/adc/adc_stm32.c:727]:
```suggestion
			data->samples_count = data->channel_count;
			data->buffer += data->channel_count;
```

I tested this PR on an STM32H745ZI-Q-NUCLEO with extra sampling enabled via `sequence->options->extra_samplings` and noticed that only the first block of samples was written, the rest of the buffer was untouched. Incrementing the buffer pointer here matches what is done in `adc_stm32_isr()` and it made `extra_samplings` write to the whole buffer.

File: drivers/adc/adc_stm32.c
Comment (heinwessels) [reply drivers/adc/adc_stm32.c]: You are indeed correct. I was looking at the documentation and not the function body. I also verified that it still works by offsetting the buffers by one, and all tests still passed. So I'm removing this check.

File: drivers/adc/adc_stm32.c
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:725]: Nice catch! Also fixed the unit tests, because it should've failed with this missing.

**Comment by heinwessels** [comment]
I added a commit to fix the `adc_dma` tests because it didn't verify that the sampled buffer is actually correct. It does this by placing a known value in the buffer beforehand, and then afterwards verifies the expected slots were overwritten.

This also led to a fix in the stm32 driver where the cache needs to be be clean _before_ a read sequence to ensure that when the cache is invalidated in the callback no data is lost.

File: drivers/adc/adc_stm32.c
Comment (erwango) [inline drivers/adc/adc_stm32.c:322]: Hopefully can be removed if using RAM_NOCACHE on SRAM2

File: tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay
Comment (erwango) [inline tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay:7]: Please use helping macros defined [here](https://github.com/zephyrproject-rtos/zephyr/pull/53486)

File: tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay
Comment (erwango) [inline tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay:24]: copy/paste ?

File: drivers/adc/adc_stm32.c
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:322]: I'm not sure if this is the right approach. It would mean I should probably add a check in the dma driver to verify that the given source and destination addresses are both in nocache regions, and fail if it's not. That might break existing setups that is using cache maintenance operations.

The reason I think it's okay for the BDMA in #53184 is because the SRAM4 it's a dedicated smallish section, that's typically not used. However, an user might see `NOCACHE` is required, then add the flag to the main SRAM section used by the software, and then loose the performance benifets of having cache.

Also, the description for [CONFIG_NOCACHE_MEMORY](https://docs.zephyrproject.org/2.7.0/reference/kconfig/CONFIG_NOCACHE_MEMORY.html) seems to lean towards attempting to use regular cache maintenance operations before adding the NOCACHE flag to an entire SRAM region.
```
This memory section can be used to perform DMA
transfers when cache coherence issues are not optimal or can not
be solved using cache maintenance operations
```

It would still be useful and workable though. In our application I would put both our ADC1 and ADC3 buffers inside SRAM4, which isn't used by the software anyway.

What do you think?

Comment (erwango) [reply drivers/adc/adc_stm32.c:322]: Right, I forgot this change isn't tightly linked to BDMA or H7. On other series, this would not apply.

Optionally, we can think of a Kconfig property (/dts ?) which would allow user to chose one of both approaches.

Comment (heinwessels) [reply drivers/adc/adc_stm32.c:322]: I can add a kconfig setting to `adc/Kconfig.stm32`, and let it only be available if the device has cache, and dma is enabled, etc. Something like:
```kconfig
if ADC_STM32_DMA
config ADC_STM32_DMA_ACTIVE_CACHE_HANDLING
	bool "STM32 ADC DMA Active Cache Handling"
        default n
	help
	  If enabled the ADC driver will do manual cache maintainence
          to the output buffer which means the buffer does not need to
          be located in a NOCACHE memory region. If not enabled the
          driver will not allow buffers not located in a NOCACHE memory
          region.
endif
```

It could be added to the dma kconfig, but I think the responsibility of handling the cache is of the thing using the dma, not the dma itself.

PS. Do you know of a way to check if an address is in a NOCACHE memory region during runtime? I'm not sure how to do that. I also asked on Discord in the `#DMA` channel, maybe we can chat there?

Comment (erwango) [reply drivers/adc/adc_stm32.c:322]: > It could be added to the dma kconfig, but I think the responsibility of handling the cache is of the thing using the dma, not the dma itself.

Looks fine to me

> Do you know of a way to check if an address is in a NOCACHE memory region during runtime?

Nope

File: tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay
Comment (heinwessels) [reply tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay]: Yup. Removed.

File: tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay
Comment (heinwessels) [reply tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay]: Done. Much nicer now! :D

File: drivers/adc/adc_stm32.c
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:322]: It doesn't seem possible with the current Zephyr API, or easily implementable (by me at least), to determine if a given address in RAM is non-cachable. Therefore, I'm not quite sure where to go from here.

Maybe we can keep this as-is for now?

Comment (heinwessels) [reply drivers/adc/adc_stm32.c:322]: After a conversation with erwango on Discord it was decided to remove the manual cache maintenance. Rather the driver will expect the buffer to be in a non-cachable SRAM region. This will be enfored by doing runtime check on the buffer supplied when a read is started, but only for the `STM32H7`. This check will only pass if the buffer is placed in a DTS SRAM region which is marked as `NON_CACHEABLE`, meaning the `.nocache` section will not work. It seems like using DTS section is the intended anyway.

Comment (heinwessels) [reply drivers/adc/adc_stm32.c]: Updated as described in previous comment.

**Comment by heinwessels** [comment]
I added an addtional commit to remove `CONFIG_TEST_USERSPACE` from the `adc_dma` tests, because I don't think it's compatible with having buffers in memory which is not defined by `ZTEST_BMEM`. And it doesn't seem required either.

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Comment by erwango** [comment]
DNM until @FRASTM could have a chance to review (probably today)

**Review by FRASTM — APPROVED** [review]

**Pull request merged**
